### PR TITLE
Implement a basic "confidence" system for fingerprinting

### PR DESCRIPTION
**Pull request by BelleNottelling**
~~I've found a situation where the current method is causing issues, however it's not my instance so I can't provide easy testing.
This draft PR is up so I can send the changes required to the user & get feedback while I await to see if it improves the situation at all.~~ I have confirmation it fixed the issue

If you want an in-depth description of the new behavior, please check the comments I added to the code which explains as they are in-depth and fully cover everything.

The TLDR of it is that this changes the failure threshold from a fixed `50%` to being `50%` when we can be confident, but the it scales up as we have less available data-points, which makes it harder for the check to fail when there's not enough info.

### CODE REVIEW

**Comment by BelleNottelling** [comment]
The fix has been confirmed to be working

**Review by jaapmarcus — APPROVED** [review]

File: src/library/FOSSBilling/Fingerprint.php [hunk 1]
```diff
@@ -130,9 +132,34 @@ public function checkFingerprint(array $fingerprint): bool
...
+         * 
+         * In the event that less that 70% of the possible values are in the fingerprint, we use the percentage off we are to calculate a higher percentage before failure.
+         * For example:
+         *  If we have 13 possible fingerprint properties and only 9 are available, that's only 69% of the possible properties and the failure threshold will be calculated as follows.
+         *  Negative weight: (1 - 0.69) / 1.25 = `0.24`
+         *  Failure threshold: 0.5 + 0.24 = `0.74`
+         *  
+         *  So in this example, the percentage wrong needs to be at or above 74% (nearly 75%) before the session is declared invalid.
+         *  If there's only 6 of 13 available that moves to 93% and at 5 of 13 it's 99%.
+         *  By doing this, we can effectively give additional headroom in situations where we are less-confident than we'd like to be and effectively completely disable the check in a worst-case situation.
```
Comment (Anuril) [inline src/library/FOSSBilling/Fingerprint.php:147]: So if I craft a request to remove the HTTP Headers REMOTE_ADDR, HTTP_REFERER, HTTP_X_FORWARDED_FOR, HTTP_ACCEPT_LANGUAGE, HTTP_ACCEPT_ENCODING, HTTP_UPGRADE_INSECURE_REQUESTS and HTTP_CF_IPCOUNTRY, I've removed 7 of 13 Values, meaning there's only 6 left to compare. Because 93% of the remaining ones need to be false before the session gets invalidated the remaining values are: userAgent, browser, browserVersion, os, HTTP_SEC_CH_UA_PLATFORM and HTTP_SEC_CH_UA_MOBILE. 

So if I get os and browser right, I can still hijack the session, right? 

Windows has 70-75% market share, Macos ~16%. Browsers are 63% Chrome and ~20% Safari.

I think we should find more and different markers, or we need to bring the agressiveness of this additional headroom down.

**Comment by Anuril** [comment]
Some reverse proxy configurations will always add HTTP_UPGRADE_INSECURE_REQUESTS to the request headers, so that won't ever change and HTTP_CF_IPCOUNTRY isn't useful if the system isn't behind cloudflare, so we're effectively down to 11 markers.

https://github.com/FOSSBilling/FOSSBilling/blob/9ce61b2aed7064df36dc24f78509c661a5072edc/src/library/FOSSBilling/Fingerprint.php#L67C56-L67C56

**Comment by BelleNottelling** [comment]
> Some reverse proxy configurations will always add HTTP_UPGRADE_INSECURE_REQUESTS to the request headers, so that won't ever change and HTTP_CF_IPCOUNTRY isn't useful if the system isn't behind cloudflare, so we're effectively down to 11 markers.

It would be great if you actually stated what your point is with this.

But besides that, the actual properties of the fingerprint are outside of what this PR changes. 

The purpose of the PR is to change the behavior of it so that it's more lenient when we have less properties available to work with so that it won't cause issues under situations that are less-than-ideal.

**Comment by Anuril** [comment]
> It would be great if you actually stated what your point is with this.

The point is that from 13 Markers, 1 is only present if the instance is protected by cloudflare, and 1 is likely not to change in setups with reverse proxies.

The result is, that only 12 markers are available to being compared, and 1 marker would always be "good" if it's behind a reverse proxy that configures HTTP_UPGRADE_INSECURE_REQUESTS.

**Review by Anuril — COMMENTED** [review]
It would've been great to submit the review before writing the other comment.

File: src/library/FOSSBilling/Fingerprint.php (hunk 1 above)
Comment (BelleNottelling) [reply src/library/FOSSBilling/Fingerprint.php:147]: If you removed `REMOTE_ADDR`, `HTTP_REFERER`, `HTTP_X_FORWARDED_FOR`, `HTTP_ACCEPT_LANGUAGE`, `HTTP_ACCEPT_ENCODING`, `HTTP_UPGRADE_INSECURE_REQUESTS`, and `HTTP_CF_IPCOUNTRY` from the header when they were initially there that would instantly be considered to be 100% wrong because the added up weights from those are 13.

Comment (BelleNottelling) [reply src/library/FOSSBilling/Fingerprint.php:147]: Any and all properties that are in one and not in the other are counted as wrong.
Any and all properties that do not match are also counted as wrong.

Your example also assumes that their IP address correctly matches the original one when it won't and there is also a fairly decent chance that the browser version will differ, which is also counted as wrong. 

Adding these two changes on top of the example you gave brings the weighted wrong percentage to 120%

**Review by Anuril — APPROVED** [review]
I'll approve until I have a better idea or more fingerprint markers.
